cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java

```diff
@@ -0,0 +1,157 @@
+import io.cdap.cdap.proto.id.ProgramRunId;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import io.cdap.cdap.spi.data.transaction.TransactionRunners;
+import io.netty.handler.codec.http.HttpRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A {@link RuntimeRequestValidator} implementation that reads from the runtime table directly.
+ */
+public final class DirectRuntimeRequestValidator implements RuntimeRequestValidator {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DirectRuntimeRequestValidator.class);
+  private static final Gson GSON = new Gson();
+
+  private final TransactionRunner txRunner;
+  private final ProgramRunRecordFetcher runRecordFetcher;
+  private final LoadingCache<ProgramRunId, Boolean> programRunsCache;
+
+  @Inject
+  DirectRuntimeRequestValidator(CConfiguration cConf, TransactionRunner txRunner,
+                                ProgramRunRecordFetcher runRecordFetcher) {
+    this.txRunner = txRunner;
+    this.runRecordFetcher = runRecordFetcher;
+
+    // Configure the cache with expiry the poll time.
+    // This helps reducing the actual lookup for a burst of requests within one poll interval,
+    // but not to keep it too long so that data becomes stale.
+    long pollTimeMillis = cConf.getLong(Constants.RuntimeMonitor.POLL_TIME_MS);
+    this.programRunsCache = CacheBuilder.newBuilder()
+      .expireAfterWrite(pollTimeMillis, TimeUnit.MILLISECONDS)
+      .build(new CacheLoader<ProgramRunId, Boolean>() {
+        @Override
+        public Boolean load(ProgramRunId programRunId) throws IOException {
+          return isValid(programRunId);
+        }
+      });
+  }
+
+  @Override
+  public void validate(ProgramRunId programRunId, HttpRequest request) throws BadRequestException {
+    boolean exists;
+    try {
+      exists = programRunsCache.get(programRunId);
+    } catch (Exception e) {
+      throw new ServiceUnavailableException(Constants.Service.RUNTIME, e);
+    }
+    if (!exists) {
+      throw new BadRequestException("Program run " + programRunId + " is not valid");
+    }
+  }
+
+  /**
+   * Checks if the given {@link ProgramRunId} is valid.
+   */
+  private boolean isValid(ProgramRunId programRunId) throws IOException {
+    RunRecordDetail runRecord = TransactionRunners.run(txRunner, context -> {
+      return AppMetadataStore.create(context).getRun(programRunId);
+    }, IOException.class);
+
+    if (runRecord != null) {
+      return !runRecord.getStatus().isEndState();
+    }
+    // If it is not found in the local store, which should be very rare, try to fetch the run record remotely.
+    try {
+      LOG.info("Fetching program run details for {} remotely", programRunId);
+      runRecord = runRecordFetcher.getRunRecordMeta(programRunId);
+      // Try to update the local store
+      insertRunRecord(programRunId, runRecord);
```

Comment:
Separate this part out means the caller to this method to do it?? But the information about that it is fetching remotely doesn't go out of this method.

Reply:
Would it be a good idea to throw `NotFoundException` in isValid when program is not present and do insertion logic in the caller.

Reply:
No. It is intention to populate the cache with a false so that we don't need to keep making remote call in case someone is calling this with a program run that is not runnning.

Reply:
I see.